### PR TITLE
chrome-remote-desktop: init at unstable-2022-02-03

### PR DESCRIPTION
**Pull request by sepiabrown**
###### Motivation for this change
#34084

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sepiabrown** [comment]
The code works well with `nixpkgs.url = "github:nixos/nixpkgs/nixos-21.05"` and `nixos-21.11` in my `flake.nix` but not with `nixos-unstable`. 

It was due to recent change in `linux-pam` (#153104). Many other applications were affected by the change. I've checked that `chrome-remote-desktop` works normally with the fix #156974.

**Comment by sepiabrown** [comment]
@jtojnar Finally! All checks are passed! Thank you for your guidance!

**Comment by PaulGrandperrin** [comment]
Hi, thanks for your work!
I tried the PR but I get this issue:
```
systemctl status chrome-remote-desktop@paulg.service
× chrome-remote-desktop@paulg.service - Chrome Remote Desktop instance for paulg
     Loaded: loaded (/etc/systemd/system/chrome-remote-desktop@paulg.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Tue 2022-02-08 22:14:08 CET; 2min 11s ago
    Process: 55099 ExecStart=/nix/store/y1bvyjldk1z6ra66bg3xsdskb3sylcz7-chrome-remote-desktop-unstable-2022-02-03/opt/google/chrome-remote-desktop/chrome-remote-desktop --start --new-session (code=exited, statu>
   Main PID: 55099 (code=exited, status=1/FAILURE)
         IP: 0B in, 0B out
        CPU: 913ms

Feb 08 22:14:07 MacBookPaul systemd[1]: Started Chrome Remote Desktop instance for paulg.
Feb 08 22:14:07 MacBookPaul systemd[55099]: pam_unix(chrome-remote-desktop:session): session opened for user paulg(uid=1000) by (uid=0)
Feb 08 22:14:08 MacBookPaul chrome-remote-desktop[55099]: 2022-02-08 22:14:08,486:WARNING:WARNING: This system uses GDM. Some GDM versions have a bug that prevents local login while Chrome Remote Desktop is runn>
Feb 08 22:14:08 MacBookPaul chrome-remote-desktop[55099]: Failed to load config: [Errno 2] No such file or directory: '/home/paulg/.config/chrome-remote-desktop/host#5ad3bb63a810807084bf8809c1f70b2d.json'
Feb 08 22:14:08 MacBookPaul chrome-remote-desktop[55099]: 2022-02-08 22:14:08,486:INFO:Cleanup.
Feb 08 22:14:08 MacBookPaul chrome-remote-desktop[55099]: WARNING: This system uses GDM. Some GDM versions have a bug that prevents local login while Chrome Remote Desktop is running. If you run into this issue,>
Feb 08 22:14:08 MacBookPaul systemd[1]: chrome-remote-desktop@paulg.service: Main process exited, code=exited, status=1/FAILURE
Feb 08 22:14:08 MacBookPaul systemd[55100]: pam_unix(chrome-remote-desktop:session): session closed for user paulg
Feb 08 22:14:08 MacBookPaul systemd[1]: chrome-remote-desktop@paulg.service: Failed with result 'exit-code'.
```

I created `~/.config/chrome-remote-desktop` manually but it doesn't change anything.

**Comment by sepiabrown** [comment]
@PaulGrandperrin
`Feb 08 22:14:08 MacBookPaul chrome-remote-desktop[55099]: 2022-02-08 22:14:08,486:WARNING:WARNING: This system uses GDM. Some GDM versions have a bug that prevents local login while Chrome Remote Desktop is runn>`
This error comes from the flie `opt/google/chrome-remote-desktop/chrome-remote-desktop` line 1687. (You can inspect it yourself by downloading `https://dl.google.com/linux/direct/chrome-remote-desktop_current_amd64.deb` and `dpkg -x chrome-remote-desktop_current_amd64.deb .`; You can also inspect it at [google git](https://chromium.googlesource.com/chromium/src/+/45fe43dc7b5e669bd71aac0db80d16a6e0e2954f/remoting/host/linux/linux_me2me_host.py#1611).

The file suggests referring to [here](https://gitlab.gnome.org/GNOME/gdm/-/issues/580) which says that login fails if user has an existing graphical session such as chrome-remote-desktop.

If you want to implement it at your own risk, you may test it by adding `substituteInPlace $out/$replacePrefix/chrome-remote-desktop --replace "if display_manager_is_gdm():${"\n"}    return False" ""` in the `patchPhase` of `default.nix`. I added the file in my [`chrome-remote-desktop-GDM`](https://github.com/sepiabrown/nixpkgs/blob/462998ebb18df09a8819d5405e9abecbd01f089e/pkgs/applications/networking/remote/chrome-remote-desktop/default.nix#L49) branch.

**Comment by elatov** [comment]
it looks like the new version of the deb package depends on a new library, see this on my machine, I had to update the hash as well:

```
> nix-prefetch-url https://dl.google.com/linux/direct/chrome-remote-desktop_current_amd64.deb --type sha256
[16.0 MiB DL]
path is '/nix/store/l38zyksq03qkba7m44k7d3gxlvd7g636-chrome-remote-desktop_current_amd64.deb'
0ci21445azn6hlqwkf8xp7q8z1r5bzl9160icilg2vplwvd39f1c
```

And I am getting this error:

```/nix/store/30h52wy2z50dq4lwnvly90nq1kwlig88-chrome-remote-desktop-unstable-2022-02-03/opt/google/chrome-remote-desktop/chrome-remote-desktop-host: error while loading shared libraries: libremoting_core.so: cannot open shared object file: No such file or directory```

**Comment by GGG-KILLER** [comment]
Just tried this on my machine and ran into the following error when the systemd service was starting up:
```
systemd[1]: Started Chrome Remote Desktop instance for ggg.
systemd[1020805]: pam_unix(chrome-remote-desktop:session): session opened for user ggg(uid=1000) by (uid=0)
chrome-remote-desktop[1020805]:   File "/nix/store/sjd4yhv1gb37bpy5ggpa7ka1xnxwn29j-chrome-remote-desktop-unstable-2022-02-03/opt/google/chrome-remote-desktop/chrome-remote-desktop", line 528
chrome-remote-desktop[1020805]:     self.launch_desktop_session()
chrome-remote-desktop[1020805]: IndentationError: unexpected indent
systemd[1]: chrome-remote-desktop@ggg.service: Main process exited, code=exited, status=1/FAILURE
systemd[1020822]: pam_unix(chrome-remote-desktop:session): session closed for user ggg
systemd[1]: chrome-remote-desktop@ggg.service: Failed with result 'exit-code'.
```

**Comment by SuperSandro2000** [comment]
Closing due to no answer from the author @sepiabrown in almost a year.

File: pkgs/applications/networking/remote/chrome-remote-desktop/default.nix [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      libdrm
+      libxkbcommon
+      wayland
+    ];
+  in ''
+        for i in $out/$replacePrefix/{chrome-remote-desktop-host,start-host,native-messaging-host,remote-assistance-host,user-session}; do
+          sed -i "s|$replacePrefix|$replaceTarget|g" $i
+          patchelf --set-rpath "${libPath}" $i
+          patchelf --set-interpreter ${glibc}/lib/ld-linux-x86-64.so.2 $i
+        done
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/remote/chrome-remote-desktop/default.nix:94]: We intend with 2 spaces in nix files.

File: pkgs/applications/networking/remote/chrome-remote-desktop/default.nix [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  ''
+    sed \
+    -e '/^.*sudo_command =/ s/"gksudo .*"/"pkexec"/' \
+    -e '/^.*command =/ s/s -- sh -c/s sh -c/' \
+    -i $out/opt/google/chrome-remote-desktop/chrome-remote-desktop
+    substituteInPlace $out/etc/opt/chrome/native-messaging-hosts/com.google.chrome.remote_desktop.json --replace $replacePrefix/native-messaging-host $out/$replacePrefix/native-messaging-host
+    substituteInPlace $out/etc/opt/chrome/native-messaging-hosts/com.google.chrome.remote_assistance.json --replace $replacePrefix/remote-assistance-host $out/$replacePrefix/remote-assistance-host
+
+    substituteInPlace $out/$replacePrefix/chrome-remote-desktop --replace "USER_SESSION_PATH = " "USER_SESSION_PATH = \"/run/wrappers/bin/crd-user-session\" #"
+    substituteInPlace $out/$replacePrefix/chrome-remote-desktop --replace /usr/bin/python3 ${python3.withPackages (ps: with ps; [ psutil ])}/bin/python3
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/remote/chrome-remote-desktop/default.nix:39]: substituteInPlace can take multiple replace arguemnts.

```suggestion
    substituteInPlace $out/$replacePrefix/chrome-remote-desktop \
      --replace "USER_SESSION_PATH = " "USER_SESSION_PATH = \"/run/wrappers/bin/crd-user-session\" #" \
      --replace /usr/bin/python3 ${python3.withPackages (ps: with ps; [ psutil ])}/bin/python3
```
and so on

File: pkgs/applications/networking/remote/chrome-remote-desktop/default.nix [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  installPhase = ''
+    mkdir $out/bin
+    makeWrapper $out/opt/google/chrome-remote-desktop/chrome-remote-desktop $out/bin/chrome-remote-desktop
+  '';
+
+  replacePrefix = "/opt/google/chrome-remote-desktop";
+  replaceTarget = "/run/current-system/sw/bin/./././";
+
+  patchPhase =
+  ''
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/remote/chrome-remote-desktop/default.nix:30]:
```suggestion
  postPatch = ''
```

File: pkgs/applications/networking/remote/chrome-remote-desktop/default.nix [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    ${dpkg}/bin/dpkg -x $src $out
+  '';
+
+  installPhase = ''
+    mkdir $out/bin
+    makeWrapper $out/opt/google/chrome-remote-desktop/chrome-remote-desktop $out/bin/chrome-remote-desktop
+  '';
+
+  replacePrefix = "/opt/google/chrome-remote-desktop";
+  replaceTarget = "/run/current-system/sw/bin/./././";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/remote/chrome-remote-desktop/default.nix:27]: Mangling nix variable to bash ones is not that great.

File: pkgs/applications/networking/remote/chrome-remote-desktop/default.nix [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+stdenv.mkDerivation rec {
+  pname = "chrome-remote-desktop";
+  version = "unstable-2022-02-03";
+
+  src = fetchurl {
+    sha256 = "sha256-CNE3kvAj7Jp4cB5x2D/YUA1H9Ri397C6rxQBRmstz1c=";
+    url = "https://dl.google.com/linux/direct/chrome-remote-desktop_current_amd64.deb";
+  };
+
+  buildInputs = [ pkgs.makeWrapper ];
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/remote/chrome-remote-desktop/default.nix:12]:
```suggestion
  nativeBuildInputs = [ makeWrapper ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**